problem-solver/cxx/format-translators-module/agent/visualisation-translator-agent/scg_visualisation_translator_agent.cpp

```diff
@@ -0,0 +1,531 @@
+    if (m_context.CheckConnector(identifiersLanguage, link, ScType::ConstPermPosArc))
+      m_context.GetLinkContent(link, identifier);
+  }
+  if (identifier.empty())
+    identifier = m_context.GetElementSystemIdentifier(elementAddr);
+  if (!identifier.empty())
+    element->setIdentifier(identifier);
+  element->setIdentifierPosition(IdentifierPosition::BOTTOM_RIGHT);
+}
+
+std::shared_ptr<Connector> SCgVisualisationTranslatorAgent::createConnector(
+    ScAddr const & connector,
+    std::shared_ptr<Element> const & baseElement,
+    std::shared_ptr<Node> const & otherElement,
+    bool const isReversed) const
+{
+  auto const & connectorElement = std::make_shared<Connector>();
+  connectorElement->setScAddress(connector);
+  connectorElement->setScType(m_context.GetElementType(connector));
+
+  setSystemIdentifierIfExists(connector, connectorElement);
+
+  connectorElement->setBaseElement(baseElement);
+  connectorElement->setOtherElement(otherElement);
+  connectorElement->setIsReversed(isReversed);
+
+  return connectorElement;
+}
+
+void SCgVisualisationTranslatorAgent::addVerticalConnector(
+    ScAddr const & otherAddr,
+    ScAddr const & connectorAddr,
+    std::shared_ptr<Connector> const & baseConnector,
+    bool const isReversed) const
+{
+  auto const otherElement = createNode(otherAddr, m_context.GetElementType(otherAddr));
+  auto const & connectorElement = createConnector(connectorAddr, baseConnector, otherElement, isReversed);
+  baseConnector->addConnector(connectorElement);
+}
+
+std::list<std::shared_ptr<Node>> SCgVisualisationTranslatorAgent::assignXCoordinates(
+    std::shared_ptr<Node> const & rootElement)
+{
+  float previousLevelMaxLinkContentSize = rootElement->getContent().size();
+  if (previousLevelMaxLinkContentSize > MAX_LINK_WIDTH)
+    previousLevelMaxLinkContentSize = MAX_LINK_WIDTH;
+  float nextLevelIndent = MIN_X;
+  float const scaleX = 6;
+  rootElement->setX(nextLevelIndent);
+  auto treeRoots = std::list{rootElement};
+
+  for (auto const & treeRoot : treeRoots)
+  {
+    treeRoot->setX(nextLevelIndent);
+    auto currentLevelRoots = std::list{treeRoot};
+    while (!currentLevelRoots.empty())
+    {
+      std::list<std::shared_ptr<Connector>> connectorsToNextLevelRoots;
+      std::list<std::shared_ptr<Node>> nextLevelRoots;
+      float maxRelationIdentifierLength = 0;
+      float maxLevelElementIdentifierLength = 0;
+      calculateNextLevelElementsAndIdentifiersLengths(
+          currentLevelRoots,
+          connectorsToNextLevelRoots,
+          nextLevelRoots,
+          maxRelationIdentifierLength,
+          maxLevelElementIdentifierLength);
+      float const indentCausedByLinkContent =
+          calculateIndentCausedByLinkContent(connectorsToNextLevelRoots, previousLevelMaxLinkContentSize);
+      float const indentDifference =
+          (std::max(maxLevelElementIdentifierLength, maxRelationIdentifierLength) + indentCausedByLinkContent + 8)
```

Comment:
What is 8 here?

Reply:
8 is used to create a little of empty space between identifier and next level elements